### DISCUSSION TITLE
AttributeError: 'ClassManager' object has no attribute 'obj' when listening for "first_init" event

### DISCUSSION THREAD

**Discussion by gottogethelp** [sqlalchemy/sqlalchemy#10499] [Usage Questions] [answered]
I'd like to update a table in a standard way when any of a variety of scripts I've written runs. To that end I'd like to use the `event.listens_for` decorator so I don't need to call my update function at the beginning of each script.

I opted to listen for the `first_init` event using the example in the SQLALchemy docs [here][1].

However, the following MRE:
```
import sqlalchemy as sa
from sqlalchemy.orm import DeclarativeBase, Mapped, mapped_column
from sqlalchemy.schema import CreateSchema, DropSchema

from database.schemas.base import Base


class Base(DeclarativeBase):
    pass


class Country(Base):
    __tablename__ = "country"
    __table_args__ = {"schema": "country_test"}

    country_id: Mapped[int] = mapped_column(sa.SmallInteger, primary_key=True)
    country_name: Mapped[str] = mapped_column(sa.String(50), unique=True)

    def __init__(self, country_name: str) -> None:
        self.country_name = country_name


@sa.event.listens_for(Country, "first_init")
def my_update_function(manager, cls):
    # code that scrapes an up to date list of countries and adds new ones as required
    print("my_update_function ran")


if __name__ == "__main__":
    engine = sa.create_engine(<conn_str>)
    connection = engine.connect()
    connection.execute(DropSchema("country_test", if_exists=True))
    connection.execute(CreateSchema("country_test"))
    Base.metadata.create_all(bind=engine)

    record_1 = Country("test1")
    record_2 = Country("test2")
```
Gives me the error message:
```
AttributeError: 'ClassManager' object has no attribute 'obj'
```
If I remove `my_update_function` then everything runs fine so I'm pretty sure the issue is with my implementation of the decorator. However, I'm not sure what the error message means so I'm a little stuck on where to go from here...

  [1]: https://docs.sqlalchemy.org/en/20/orm/events.html#sqlalchemy.orm.InstanceEvents.first_init

**Comment by zzzeek** [comment, marked as answer]
I'm sorry, this event seems to be entirely untested and has likely not worked for years.   issue #10500 is added.   


you might want to use a more common event hook like [mapper_configured](https://docs.sqlalchemy.org/en/20/orm/events.html#sqlalchemy.orm.MapperEvents.mapper_configured) that is more aligned to when a class is mapped.
